## CNV analysis

Copy number variation usually occurs in blood tumor cells. [CopyKAT](https://github.com/navinlabcode/copykat) is used to analyze CNV. The inferred tumor area and normal area are shown in the following figure.

<center>
    <img src="..\\Step7_CNV_analysis\\png\\CNV_Spatial.png" width="40%"/>
</center>

Here is the generated heatmap plot for estimated copy numbers. Rows are single cells; columns are 220kb bins in genomic order.

<center>
    <img src="..\\Step7_CNV_analysis\\png\\CNV_heatmap.png" width="70%"/>
</center>

<!-- The visualization of two copy number subpopulations of aneuploidy tumor cells is as follows. -->

<!-- <center> -->
<!--     <img src="..\\Step7_CNV_analysis\\png\\Tumor_subpopulations.png" width="70%"/> -->
<!-- </center> -->

The relevant argument settings are as follows.

| Arguments | value | &nbsp;&nbsp;&nbsp;&nbsp; Meaning &nbsp;&nbsp;&nbsp;&nbsp; |
| :---: | :---: | :---: |
| `copykat.genome` | `r copykat.genome` | 'hg20' or 'mm10' |
| `copykat.LOW.DR` | `r copykat.LOW.DR` | Minimal population fractions of genes for smoothing |
| `copykat.UP.DR` | `r copykat.UP.DR` | Minimal population fractions of genes for segmentation |
| `copykat.win.size` | `r copykat.win.size` | Minimal window sizes for segmentation |
| `copykat.distance` | `r copykat.distance` | Distance methods include 'euclidean', and correlation converted distance include 'pearson' and 'spearman' |
| `copykat.n.cores` | `r copykat.n.cores` | Number of cores for parallel computing |

