## Deconvolution

Deconvolution of spatial transcriptome data can identify the positions of different cell types in space. [cell2location](https://github.com/BayraktarLab/cell2location) is used to map single cell RNA-seq data to spatial transcriptomics and establish a spatial mapping of cell types. The abundance of each type of cells in spatial coordinates is shown as follows. 

<center>
    <img src="..\\Step8_Deconvolution\\png\\Cell2loc_total.png" width="100%" alt="The absolute abundance of each cell type"/>
</center>
<center style="font-size:14px;color:#C0C0C0;text-decoration:underline">The absolute abundance of each cell type</center>

<!-- <center> -->
<!--     <img src="..\\Step8_Deconvolution\\png\\Cell2loc_total_norm.png" width="100%" alt="The relative abundance of each cell type"/> -->
<!-- </center> -->
<!-- <center style="font-size:14px;color:#C0C0C0;text-decoration:underline">The relative abundance of each cell type</center> -->

<!-- where the calculation method for relative abundance is -->
<!-- $$abundance_{relative} = \frac{abundance_{absolute}}{nUMI} * median(nUMI)$$ -->

The relevant argument settings are as follows.

| Arguments | value | &nbsp;&nbsp;&nbsp;&nbsp; Meaning &nbsp;&nbsp;&nbsp;&nbsp; |
| :---: | :---: | :---: |
| `cell2loc.sc.h5ad.dir` | `r cell2loc.sc.h5ad.dir` | Path to scRNA-seq h5ad file |
| `cell2loc.sc.max.epoch` | `r cell2loc.sc.max.epoch` | Number of maximum epochs to train the reference model |
| `cell2loc.st.max.epoch` | `r cell2loc.st.max.epoch` | Number of maximum epochs to train the mapping model |
| `cell2loc.use.gpu` | `r cell2loc.use.gpu` | Whether to use GPU |
